src/main.cc

```diff
@@ -521,8 +521,9 @@ mainHandleCommandLineOption(const int optId, const char *optValue)
 
     case 'd':
         /** \par d
-         * Set global option Debug::log_stderr to the number given following the option */
-        Debug::log_stderr = xatoi(optValue);
+         * debugs() messages with the given debugging level (and the more
+         * important ones) should be written to stderr */
+        Debug::ResetStderrLevel(xatoi(optValue));
```

Comment:
Missing check that -X is not forcing the value to -d9.

 And -X handling missing check for channels already activated that need a reset to level-9.

Reply:
This one is important please. I recall a number of times users have provided incorrect or useless debug traces because of these options behaviour colliding invisibly.

If you disagree on the exact interaction behaviour I outlined then at least add a clear debugs() to tell users what level is actually going to be logged when these values differ.

Reply:
> I recall a number of times users have provided incorrect or useless debug traces because of these options behaviour colliding invisibly.

I doubt we can make this complex issue (three individually configurable channels, 9 levels, numerous debugging sections) disappear completely, but I hope that our changes will reduce the number of cases where users get not what they should expect (and cannot figure it out on their own). If needed, I can analyze any specific examples you have in mind (i.e. which options you asked users to use, which options they actually used, and what was wrong with the collected output).

> If you disagree on the exact interaction behaviour I outlined then at least add a clear debugs() to tell users what level is actually going to be logged when these values differ.

Yes, I disagree (as discussed [earlier](https://github.com/squid-cache/squid/pull/950#issuecomment-994289167
)). I added a debugs() for admins using `-X`. Please see 50d363b. I hope it delivers what you are looking for despite my incomplete understanding of what specific problem you are trying to address.

----

And, in case you were also waiting for explicit responses to the original change request, here they are:

> Missing check that -X is not forcing the value to -d9.
> And -X handling missing check for channels already activated that need a reset to level-9.

`-X` unconditionally resets all _section_ levels to 9 (see the code quoted below). As for the channels:

* stderr channel: `-X` should not force `-d9` as already discussed and illustrated at https://github.com/squid-cache/squid/pull/950#issuecomment-994289167
* syslog channel: `-X` should not force "level-9" for the syslog channel because that channel never logs anything above level 1 (except for forced alerts, but those are never above level 1 IIRC).
* cache.log channel: This channel does not have a level that can be forced or even configured! Or, if you prefer, the cache.log channel level is essentially hard-coded to infinity -- that channel logs _everything_ that the global section filter passes. That global section filter is applied before _any_ channel gets control. The corner cases of early and late debugging messages aside (they are discussed [elsewhere](https://github.com/squid-cache/squid/pull/950#issuecomment-994292460)), this means that a successfully opened cache.log will receive all debug messages from its opening to its closure.


```C++
case 'X':
    /** \par X
     * Force full debugging */
    Debug::parseOptions("rotate=0 ALL,9");
    Debug::override_X = 1;
    sigusr2_handle(SIGUSR2);
    break;
```

Reply:
Thank you. I consider this resolved.